src/main/scala/is/hail/expr/types/TStructBase.scala

```diff
@@ -0,0 +1,208 @@
+package is.hail.expr.types
+
+import is.hail.annotations._
+import is.hail.asm4s.{Code, _}
+import is.hail.check.Gen
+import is.hail.utils._
+import org.apache.spark.sql.Row
+import org.json4s.jackson.JsonMethods
+
+import scala.reflect.{ClassTag, classTag}
+
+abstract class TStructBase extends Type {
+  def types: IndexedSeq[Type]
```

Comment:
this will totally destroy performance. I think that there should be an abstract `getTypes` method which returns an array, but that array should actually be stored as as final val inside of TStructBase.